shimmy/meltingpot_compatibility.py

```diff
@@ -42,41 +39,22 @@ class MeltingPotCompatibilityV0(ParallelEnv, EzPickle):
 
     def __init__(
         self,
-        substrate_name: str,
+        env,
         render_mode: str | None = None,
         max_cycles: int = MAX_CYCLES,
-        env: meltingpot.python.utils.substrates.substrate.Substrate | None = None,
     ):
         """Wrapper that converts a openspiel environment into a pettingzoo environment.
 
         Args:
-            substrate_name (str): name of meltingpot substrate to load
+            env (meltingpot.python.utils.substrates.substrate.Substrate): existing meltingpot env to use
             render_mode (Optional[str]): render_mode
             max_cycles (Optional[int]): maximum number of cycles (steps) before termination
-            env (Optional[meltingpot.python.utils.substrates.substrate.Substrate]): existing meltingpot env to use
 
         """
-        EzPickle.__init__(self, substrate_name, render_mode, max_cycles, env)
+        EzPickle.__init__(self, env, render_mode, max_cycles)
 
         self.max_cycles = max_cycles
-        if env is not None:
-            self._env = env
-        else:
-            # Create env config
```

Comment:
Can we not just make meltingpot a local import here to avoid the issue?

Reply:
> Can we not just make meltingpot a local import here to avoid the issue?
> 
> Then sphinx won't load it

I could've done that instead but it feels pretty hack-ey, feel like making a separate helper function which imports meltingpot.python there is better. I feel like refactoring it to match other envs makes sense but I could add back in the substrate name argument and if else logic to load it. Also very clunky to deal with the optional env or substrate_name arguments, cause you have to have one of them be required so if you want to convert an env that already exists you have to pass in the argument `substrate_name=""`.

Reply:
I would partially agree but `meltingpot` is not used in any other part of the environment, and for the type check you can just use `TYPE_CHECKING` therefore, I think it is simpler to just have a single file. 

> Also very clunky to deal with the optional env or substrate_name arguments, cause you have to have one of them be required so if you want to convert an env that already exists you have to pass in the argument substrate_name="".

I think this is just a mistake in the implementation, looking at the `__init__` again, I think it should be 
```python
def __init__(
        self,
        substrate_name: str | None = None,
        env: meltingpot.python.utils.substrates.substrate.Substrate | None = None,
        render_mode: str | None = None,
        max_cycles: int = MAX_CYCLES,
    ):
```

Reply:
Using an import halfway through a file is super clunky and raises pyright errors, keeping the boiler plate code specific to melting pot in the utilities is best imo. Gym you can load with one line but every other pettingzoo supported env type you need to initialize the env separately and the shimmy compatibility wrapper is then used with the existing env. 

For making the release notes/blog post it makes more sense to just say because petting zoo does not have a make function you need to initialize the env in a separate line and then wrap it with the compatibility wrapper.

Reply:
Not sure what you mean by TYPE_CHECKING though, is there a way to avoid the module not found error if it’s just for the type? I tried separating the env loading using the if else so that it wouldn’t always try to load an env from file, and sphinx still had an error that it couldn’t find the module in the helper file. The only solution I see is to make the default be passing in an existing env so that the code doesn’t try to import melting pot by default, but that’s not consistent with any of the other pettingzoo wrappers. 

Should we go through and make openspiel and DM lab also support taking an environment name in their constructor?

Reply:
> is there a way to avoid the module not found error if it’s just for the type? 

https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import meltingpot.python
```

> Using an import halfway through a file is super clunky and raises pyright errors, keeping the boiler plate code specific to melting pot in the utilities is best IMO.

I might be missing something, but if you import `meltingpot` at the top of the compatibility function or in a utility file then you still have the same issue. Therefore, the only solution is to use a local import, so you might be right but in the end it is simpler to have the local import in the same file rather than a file with a single function that doesn't need to exist imo.

Reply:
> Should we go through and make openspiel and DM lab also support taking an environment name in their constructor?

If there is a equivalent function to `build` or `make` in openspiel and dm-lab then I think yes, otherwise no

Reply:
Well the problem is it's not just for type checking, the helper functions which are used to load the config from substrate name also require meltingpot.python. If we made it not load from substrate name by default and then did a local import in that case it could work, but I still feel that bit at least makes more sense to do in a separate file because it's like 10 lines of boilerplate code at the top of the init which don't need to be there.

Still think it's bad to be inconsistent between meltingpot/openspiel/dm lab multiagent, so whatever we decide should be done for all of them (e.g., use existing env by default or load from env name optionally and do local import in that case). I can see a good argument for including the boilerplate code needed to load an environment in each of the compatibility wrappers though, so the user doesn't have to know the syntax for loading one they just need to know the name.

Reply:
Okay I've pushed a commit which I think combines both of our approaches and makes things cleaner (looking back at the code my first solution was a bit of a workaround your TYPE_CHECKING thing definitely is the correct way to handle the issues, versus just removing type hints for the init)